File: engines/go/internal/session/session.go
package session

import (
	"encoding/json"
	"io"
)

type Model struct {
	ID          string `json:"id"`
	DisplayName string `json:"display_name"`
}

type ContextWindow struct {
	UsedPercentage    float64 `json:"used_percentage"`
	ContextWindowSize float64 `json:"context_window_size"`
	TotalInputTokens  float64 `json:"total_input_tokens"`
	TotalOutputTokens float64 `json:"total_output_tokens"`
}

type Cost struct {
	TotalCostUSD       float64 `json:"total_cost_usd"`
	TotalDurationMs    float64 `json:"total_duration_ms"`
	TotalAPIDurationMs float64 `json:"total_api_duration_ms"`
	TotalLinesAdded    float64 `json:"total_lines_added"`
	TotalLinesRemoved  float64 `json:"total_lines_removed"`
}

type Workspace struct {
	ProjectDir string `json:"project_dir"`
	CurrentDir string `json:"current_dir"`
}

type Session struct {
	Cwd            string        `json:"cwd"`
	SessionID      string        `json:"session_id"`
	Model          Model         `json:"model"`
	ContextWindow  ContextWindow `json:"context_window"`
	Cost           Cost          `json:"cost"`
	Version        string        `json:"version"`
	Workspace      Workspace     `json:"workspace"`
	TranscriptPath string        `json:"transcript_path"`
}

func Parse(r io.Reader) Session {
	var s Session
	data, err := io.ReadAll(r)
	if err != nil {
		return s
	}
	_ = json.Unmarshal(data, &s)
	return s
}

File: engines/rust/src/session.rs
use serde::Deserialize;
use std::io::Read;

#[allow(dead_code)]
#[derive(Deserialize, Default)]
pub struct Model {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub display_name: String,
}

#[allow(dead_code)]
#[derive(Deserialize, Default)]
pub struct ContextWindow {
    #[serde(default)]
    pub used_percentage: f64,
    #[serde(default)]
    pub context_window_size: f64,
    #[serde(default)]
    pub total_input_tokens: f64,
    #[serde(default)]
    pub total_output_tokens: f64,
}

#[derive(Deserialize, Default)]
pub struct Cost {
    #[serde(default)]
    pub total_cost_usd: f64,
    #[serde(default)]
    pub total_duration_ms: f64,
    #[serde(default)]
    pub total_api_duration_ms: f64,
    #[serde(default)]
    pub total_lines_added: f64,
    #[serde(default)]
    pub total_lines_removed: f64,
}

#[allow(dead_code)]
#[derive(Deserialize, Default)]
pub struct Workspace {
    #[serde(default)]
    pub project_dir: String,
    #[serde(default)]
    pub current_dir: String,
}

#[allow(dead_code)]
#[derive(Deserialize, Default)]
pub struct Session {
    #[serde(default)]
    pub cwd: String,
    #[serde(default)]
    pub session_id: String,
    #[serde(default)]
    pub model: Model,
    #[serde(default)]
    pub context_window: ContextWindow,
    #[serde(default)]
    pub cost: Cost,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub workspace: Workspace,
    #[serde(default)]
    pub transcript_path: String,
}

pub fn parse(mut reader: impl Read) -> Session {
    let mut buf = Vec::new();
    if reader.read_to_end(&mut buf).is_err() {
        return Session::default();
    }
    serde_json::from_slice(&buf).unwrap_or_default()
}
